src/mono/mono/mini/ssa.c

```diff
@@ -284,49 +307,103 @@ mono_ssa_rename_vars (MonoCompile *cfg, int max_vars, MonoBasicBlock *bb, gboole
+	int i;
+
+	for (i = stack_history_len - 1; i >= 0; i--) {
+		stack [stack_history [i].idx] = stack_history [i].var;
 	}
+}
 
-	/* Rename PHI arguments in succeeding bblocks */
-	for (i = 0; i < bb->out_count; i++) {
-		MonoBasicBlock *n = bb->out_bb [i];
+typedef struct {
+	GSList *blocks;
+	RenameInfo *history;
+	int size;
+	int len;
+} RenameStackInfo;
 
-		for (j = 0; j < n->in_count; j++)
-			if (n->in_bb [j] == bb)
-				break;
-		
-		for (ins = n->code; ins; ins = ins->next) {
-			if (MONO_IS_PHI (ins)) {
-				idx = ins->inst_c0;
-				if (stack [idx])
-					new_var = stack [idx];
-				else
-					new_var = cfg->varinfo [idx];
-#ifdef DEBUG_SSA
-				printf ("FOUND PHI %d (%d, %d)\n", idx, j, new_var->inst_c0);
-#endif
-				ins->inst_phi_args [j + 1] = new_var->dreg;
-				record_use (cfg,  new_var, n, ins);				
-				if (G_UNLIKELY (cfg->verbose_level >= 4))
-					printf ("\tAdd PHI R%d <- R%d to BB%d\n", ins->dreg, new_var->dreg, n->block_num);
+/**
+ * mono_ssa_rename_vars:
+ * Implement renaming of SSA variables. Also compute def-use information in parallel.
+ * \p stack_history points to an area of memory which can be used for storing changes 
+ * made to the stack, so they can be reverted later.
+ */
+static void
+mono_ssa_rename_vars (MonoCompile *cfg, int max_vars, MonoBasicBlock *bb)
+{
+	GSList *blocks = NULL;
+	RenameStackInfo* rename_stack;
+	int rename_stack_size, rename_stack_idx = 0;
+	RenameInfo *stack_history;
+	int stack_history_size;
+	gboolean *originals;
+	guint32 *lvreg_stack;
+	gboolean *lvreg_defined;
+	MonoInst **stack;
+
+	stack = g_newa (MonoInst*, cfg->num_varinfo);
+	memset (stack, 0, sizeof (MonoInst *) * cfg->num_varinfo);
+	lvreg_stack = g_new0 (guint32, cfg->next_vreg);
+	lvreg_defined = g_new0 (gboolean, cfg->next_vreg);
+	stack_history_size = 10240;
+	stack_history = g_new (RenameInfo, stack_history_size);
+	originals = g_new0 (gboolean, cfg->num_varinfo);
+	rename_stack_size = 16;
+	rename_stack = g_new (RenameStackInfo, rename_stack_size);
+
+	do {
+		if (G_UNLIKELY (cfg->verbose_level >= 4))
+			printf ("\nRENAME VARS BLOCK %d:\n", bb->block_num);
+
+		int stack_history_len = create_new_vars (cfg, max_vars, bb, originals,  stack, lvreg_stack, lvreg_defined, &stack_history, &stack_history_size);
+		rename_phi_arguments_in_out_bbs (cfg, bb, stack);
+
+		if (bb->dominated) {
+			if (rename_stack_idx >= rename_stack_size - 1) {
+				rename_stack_size += MIN(rename_stack_size, 1024);
+				rename_stack = g_realloc(rename_stack, sizeof(RenameStackInfo)*rename_stack_size);
 			}
-			else
-				/* The phi nodes are at the beginning of the bblock */
-				break;
-		}
-	}
 
-	if (bb->dominated) {
-		for (tmp = bb->dominated; tmp; tmp = tmp->next) {
-			mono_ssa_rename_vars (cfg, max_vars, (MonoBasicBlock *)tmp->data, originals_used, stack, lvreg_stack, lvreg_defined, stack_history + stack_history_len, stack_history_size - stack_history_len);
+			RenameStackInfo* info = (RenameStackInfo*)rename_stack + rename_stack_idx;
```

Comment:
```suggestion
			RenameStackInfo* info = rename_stack [rename_stack_idx];
```

Reply:
We need pointer here. I can get rid of the unneeded cast at least.